cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -0,0 +1,419 @@
+
+      // add archive file
+      if (localFile.isArchive()) {
+        LOG.debug("Adding {} as archive.", localFileName);
+        hadoopJobBuilder.addArchiveUris(fileName);
+      }
+
+      // add jar file
+      if (localFile.getName().endsWith("jar")) {
+        LOG.info("Adding {} as jar.", localFileName);
+        hadoopJobBuilder.addJarFileUris(fileName);
+      } else {
+        // add all the other files as file
+        LOG.info("Adding {} as file.", localFileName);
+        hadoopJobBuilder.addFileUris(fileName);
+      }
+    }
+
+    return SubmitJobRequest.newBuilder()
+      .setRegion(region)
+      .setProjectId(projectId)
+      .setJob(Job.newBuilder()
+                // use program run uuid as hadoop job id on dataproc
+                .setReference(JobReference.newBuilder().setJobId(runInfo.getRun()))
+                // place the job on provisioned cluster
+                .setPlacement(JobPlacement.newBuilder().setClusterName(clusterName).build())
+                // add same labels as provisioned cluster
+                .putAllLabels(labels)
+                .setHadoopJob(hadoopJobBuilder.build())
+                .build())
+      .build();
+  }
+
+  /**
+   * Returns {@link RuntimeJobStatus}.
+   */
+  private RuntimeJobStatus getRuntimeJobStatus(Job job) {
+    JobStatus.State state = job.getStatus().getState();
+    LOG.debug("Dataproc job {} is in state {}.", job.getReference().getJobId(), state);
+
+    RuntimeJobStatus runtimeJobStatus;
+    switch (state) {
+      case STATE_UNSPECIFIED:
+      case SETUP_DONE:
+      case PENDING:
+        runtimeJobStatus = RuntimeJobStatus.STARTING;
+        break;
+      case RUNNING:
+        runtimeJobStatus = RuntimeJobStatus.RUNNING;
+        break;
+      case DONE:
+        runtimeJobStatus = RuntimeJobStatus.COMPLETED;
+        break;
+      case CANCEL_PENDING:
+      case CANCEL_STARTED:
+        runtimeJobStatus = RuntimeJobStatus.STOPPING;
+        break;
+      case CANCELLED:
+        runtimeJobStatus = RuntimeJobStatus.STOPPED;
+        break;
+      case ERROR:
+        runtimeJobStatus = RuntimeJobStatus.FAILED;
+        break;
+      default:
+        // this needed for ATTEMPT_FAILURE state which is a state for restartable job. Currently we do not launch
+        // restartable jobs
+        throw new IllegalStateException(String.format("Unsupported job state %s of the dataproc job %s on cluster %s.",
+                                                      job.getStatus().getState(), job.getReference().getJobId(),
+                                                      job.getPlacement().getClusterName()));
+
+    }
+    return runtimeJobStatus;
+  }
+
+  /**
+   * Stops the dataproc job. Returns job object if it was stopped.
+   */
+  @Nullable
+  private Job stopJob(RuntimeJobId runtimeJobId) throws Exception {
+    String jobId = runtimeJobId.getRuntimeJobId();
+    Job job = null;
```

Comment:
This variable is not needed. Just return directly.

```
try {
  return jobControllerClient.cancelJob(...);
} catch (...) {
}
return null;

Reply:
modified it to not return job. The cleanup would happen in destroy() method.